crates/storage/provider/src/providers/mod.rs

```diff
@@ -99,8 +103,47 @@ impl<DB: Database> BlockProvider for ShareableDatabase<DB> {
         Ok(ChainInfo { best_hash, best_number, last_finalized: None, safe_finalized: None })
     }
 
-    fn block(&self, _id: BlockId) -> Result<Option<Block>> {
-        // TODO
+    fn block(&self, id: BlockId) -> Result<Option<Block>> {
+        let block_number = match id {
+            BlockId::Number(num) => self.convert_block_number(num)?,
+            BlockId::Hash(hash) => self.block_number(H256(hash.0))?,
+        };
+
+        if let Some(number) = block_number {
+            if let Some(header) = self.header_by_number(number)? {
+                let tx = self.db.tx()?;
+                let body = tx
+                    .get::<tables::BlockBodies>(header.number)?
+                    .ok_or(Error::BlockBody { number })?;
+
+                let mut tx_cursor = tx.cursor_read::<tables::Transactions>()?;
+                let mut transactions = Vec::with_capacity(body.tx_count as usize);
+                for id in body.tx_id_range() {
+                    let (_, transaction) =
+                        tx_cursor.seek_exact(id)?.ok_or(Error::Transaction { id })?;
+                    transactions.push(transaction);
+                }
+
+                let ommers = tx.get::<tables::BlockOmmers>(header.number)?.map(|o| o.ommers);
+
+                let header_timestamp = header.timestamp;
+                let mut block = Block {
+                    header,
+                    body: transactions,
+                    ommers: ommers.unwrap_or_default(),
+                    withdrawals: None,
+                };
+
+                if self.chain_spec.fork(Hardfork::Shanghai).active_at_timestamp(header_timestamp) {
```

Comment:
I don't see why we need to have this switch here, is the `None` returned by `tx.get` not enough to signify this? We can assume that at this point, if withdrawals do not exist for the block, then we are not in Shanghai

Reply:
we don't store empty withdrawals. this has encoding implications on whether withdrawals field will be omitted or an empty array

Reply:
Is there a distinction in the spec?

Reply:
@onbjerg spec doesn't cover rpc responses

Reply:
Then I would opt to not pull in chain spec for now

Reply:
i disagree, because before shanghai `withdrawals` do not exist and this might not be universal across chains

Reply:
Okay